### PR TITLE
Wasm initcontainers

### PR DESCRIPTION
**Pull request by Sodman**
(Docs only change)

Better document deploying wasm filters using initContainers, for scenarios where using WebAssembly Hub isn't an option.

### CODE REVIEW

File: docs/content/installation/advanced_configuration/wasm.md [hunk 1]
```diff
@@ -98,7 +98,183 @@ If your image isn't hosted on an image registry, such as [WebAssembly Hub](https
...
-When loading directly from file, you'll need to ensure that the given `filePath` contains your `.wasm` file. One way to do this for example, would be using an `initContainer` on your `gatewayProxy` deployment to load the `.wasm` file into a shared `volume`.
+When loading directly from file, you'll need to ensure that the given `filePath` contains your `.wasm` file. 
+
+## Loading a wasm filter image from an initContainer
+
+In some circumstances, using [WebAssembly Hub](https://webassemblyhub.io/) as your wasm filter image repository may not be possible, for example due to enterprise networking restrictions. One way to deploy a wasm filter without going through WebAssembly Hub, is to use an `initContainer` on your `gatewayProxy` deployment to load the `.wasm` file into a shared `volume`. This section will walk you through setting this up.
+
+### Prerequisites
+
+We are assuming you already have a wasm filter created and built locally. You can replicate this by running the two commands below. Alternatively, you can follow the more in-depth guide in the WebAssembly Hub docs [here](https://docs.solo.io/web-assembly-hub/latest/tutorial_code/build_tutorials/building_cpp_filters/).
```
Comment (sam-heilbron) [inline docs/content/installation/advanced_configuration/wasm.md:109]: In the `building_cpp_filters` tutorial, there are references to `Istio 1.5.x` and `Gloo 1.3.x`, those have now been updated to more recent versions. Can we update the docs as well or is there a reason that we leave those docs pinned to previous versions?

**Review by sam-heilbron — APPROVED** [review]
This looks great! I found it really easy to reason through, and felt that the added commentary explaining each step made it easy to follow why each part was necessary. Perhaps it's not necessary for a tutorial like this, but do you think there's value in a 'common errors' section or something like that? A list of common errors people see (when they've made a mistake configuring it) and the appropriate action to take, or method to resolve. Not required for this by any means, just a thought.

File: docs/content/installation/advanced_configuration/wasm.md [hunk 2]
```diff
@@ -98,7 +98,183 @@ If your image isn't hosted on an image registry, such as [WebAssembly Hub](https
...
+        terminationMessagePolicy: File
+        volumeMounts:
+        - mountPath: /etc/envoy
+          name: envoy-config
+        - mountPath: /wasm-filters
+          name: wasm-filters
+      initContainers:
+      - name: wasm-image
+        image: localhost:8888/myorg/my-wasm-getter:1.0.0
+        imagePullPolicy: Never
```
Comment (sam-heilbron) [inline docs/content/installation/advanced_configuration/wasm.md:218]: Why is the imagePullPolicy set to never?

File: docs/content/installation/advanced_configuration/wasm.md [hunk 3]
```diff
@@ -98,7 +98,183 @@ If your image isn't hosted on an image registry, such as [WebAssembly Hub](https
...
+          runAsNonRoot: true
+          runAsUser: 10101
+        terminationMessagePath: /dev/termination-log
+        terminationMessagePolicy: File
+        volumeMounts:
+        - mountPath: /etc/envoy
+          name: envoy-config
+        - mountPath: /wasm-filters
+          name: wasm-filters
+      initContainers:
```
Comment (sam-heilbron) [inline docs/content/installation/advanced_configuration/wasm.md:215]: Is there a best practice for defining multiple wasm filters? Should each be done in it's own initContainer or all collectively? Perhaps it's more of a decision by those implementing.

File: docs/content/installation/advanced_configuration/wasm.md (hunk 2 above)
Comment (Sodman) [reply docs/content/installation/advanced_configuration/wasm.md:218]: The original reason was that I was testing locally in KIND, and manually loading the image in with `kind load docker-image ...` but now that you mention it `IfNotPresent` is probably a more appropriate value for a generic case where we don't know much about the reader's environment. I'll update it!

File: docs/content/installation/advanced_configuration/wasm.md (hunk 3 above)
Comment (Sodman) [reply docs/content/installation/advanced_configuration/wasm.md:215]: There's not really an established best-practice here yet, the only thing that really matters is that all of the filters have unique paths (ie that they're not overwriting one another). One filter per [init]container would be my preference, as they can be versioned/maintained/redeployed separately. If I were doing multiple filters, I think it makes sense to put them in the same directory but give them each better unique names than `filter.wasm`. Ultimately, it's up to the implementation team to decide their approach.

File: docs/content/installation/advanced_configuration/wasm.md (hunk 1 above)
Comment (Sodman) [reply docs/content/installation/advanced_configuration/wasm.md:109]: Those docs are definitely due an update in general, as we're now up to `Istio 1.8.x` / `Gloo 1.6.x`... In this case in particular though I _believe_ the steps should be the same (other than picking the specific version!).

**Comment by Sodman** [comment]
Re: common errors, I think usually our approach to things like this in the docs has been to listen to feedback from folks reading the docs and try to first make the docs clearer to prevent user errors in the first place, or if that's not possible, put in one of those blue floating bubble <note>'s with some "FYI if you see X, you should confirm that you've configured Y correctly", or a troubleshooting section at the end. 

Was there any errors in particular you bumped into that we might want to document? Maybe we could add a note about testing the "image getter" docker container before trying it e2e with the mount? Also maybe some notes on how to confirm that the filter was loaded into Envoy by looking at the admin config dump?

The most common errors I foresee see people running into here are probably more related to specific filters not being compatible with specific runtimes, which is probably outside of the scope of this section.

**Pull request merged**